test/new_relic/agent/transaction/request_attributes_test.rb

```diff
@@ -77,6 +161,118 @@ def test_sets_method_from_request
 
           assert_equal 'POST', attrs.request_method
         end
+
+        def test_by_default_only_a_base_set_of_request_headers_are_captured
+          skip_unless_minitest5_or_above
+
+          with_config(allow_all_headers: false, :'attributes.include' => [], :'attributes.exclude' => []) do
+            attrs = RequestAttributes.new(RACK_REQUEST)
+
+            BASE_HEADERS_MAP.each do |method, definition|
+              assert_equal definition[:value],
+                attrs.send(method),
+                "Expected RequestAttributes##{method} to yield >#{definition[:value]}<, got >#{attrs.send(method)}<"
+            end
+            assert_equal NewRelic::EMPTY_HASH, attrs.other_headers, 'Did not expect to find other headers'
+          end
+        end
+
+        def test_if_allow_all_headers_is_specified_then_allow_them_all
+          skip 'Test requires Rack v2+' unless Rack.respond_to?(:release) && !Rack.release.start_with?('1.')
+          skip_unless_minitest5_or_above
+
+          with_config(allow_all_headers: true) do
+            attrs = RequestAttributes.new(RACK_REQUEST)
+
+            BASE_HEADERS_MAP.each do |method, definition|
+              assert_equal definition[:value],
+                attrs.send(method),
+                "Expected RequestAttributes##{method} to yield >#{definition[:value]}<, got >#{attrs.send(method)}<"
+            end
+
+            assert_equal attrs.other_headers.size, OTHER_HEADERS_MAP.keys.size
+
+            attrs.other_headers.each do |header, value|
+              assert_equal OTHER_HEADERS_MAP[header], value, "Expected attribute '#{header}' to have value " +
+                "'#{OTHER_HEADERS_MAP[header]}', but it had value '#{value}'"
+            end
+          end
+        end
+
+        def test_agent_attributes_on_transaction_with_default_config
+          in_transaction do |txn|
+            attrs = RequestAttributes.new(RACK_REQUEST)
+            attrs.assign_agent_attributes(txn)
+            txn_attrs = txn.attributes.agent_attributes_for(NewRelic::Agent::AttributeFilter::DST_TRANSACTION_TRACER)
+            BASE_HEADERS_MAP.values do |definition|
+              agent_attr = definition[:agent_attribute]
+              next unless agent_attr
+
+              assert_equal definition[:value],
+                txn_attrs[agent_attr],
+                "Agent attribute '#{agent_attr}' had value '#{txn_attrs[agent_attr]}' instead of " +
+                  "'#{definition[:value]}'"
+            end
+          end
+        end
+
+        def test_the_use_of_attributes_include_as_an_allowlist
+          skip_unless_minitest5_or_above
+
+          with_config(allow_all_headers: true,
+            'attributes.include': %w[request.headers.contentType],
+            'attributes.exclude': %w[request.*]) do
```

Comment:
I didn't realize you needed to use `attributes.include` and `attributes.exclude` together to create an allowlist. Cool.

Reply:
The need to use both in tandem like this depends on what the agent's defaults are for inclusion and exclusion.

It seems to be that custom specified inclusions are added to the list of the agent's default inclusions, and custom specified exclusions are added to the list of the agent's default exclusions. So to truly be able to define a complete allowlist or blocklist, the customer has to either have some knowledge of the defaults or use both config values in tandem like this.